bootstrap/kubeadm/controllers/kubeadmconfig_controller.go

```diff
@@ -195,23 +196,38 @@ func (r *KubeadmConfigReconciler) Reconcile(req ctrl.Request) (_ ctrl.Result, re
 		return ctrl.Result{}, err
 	}
 
+	// Attempt to Patch the KubeadmConfig object and status after each reconciliation if no error occurs.
+	defer func() {
+		// always update the readyCondition; the summary is represented using the "1 of x completed" notation.
+		conditions.SetSummary(config, conditions.WithStepCounter(1))
+
+		if err := patchHelper.Patch(ctx, config); err != nil {
+			log.Error(rerr, "Failed to patch config")
+			if rerr == nil {
+				rerr = err
+			}
+		}
+	}()
+
 	switch {
 	// Wait for the infrastructure to be ready.
 	case !cluster.Status.InfrastructureReady:
 		log.Info("Cluster infrastructure is not ready, waiting")
+		conditions.MarkFalse(config, bootstrapv1.DataSecretAvailableCondition, bootstrapv1.WaitingForClusterInfrastructureReason, clusterv1.ConditionSeverityInfo, "")
```

Comment:
I'm still a bit worried that we could potentially setting this condition false regardless of whether the data secret might already exist, but something we can address separately as well.

Reply:
+1, there is definitely lots of work that can be done to restructure cabpk

Reply:
100% agreed. It is tracked by https://github.com/kubernetes-sigs/cluster-api/issues/3134